moviebot/nlu/annotation/joint_bert/dataset.py

```diff
@@ -0,0 +1,161 @@
+
+    Args:
+        data: The input data.
+
+    Yields:
+        A tuple of the intent, text, and slot annotations.
+    """
+    for intent in data.keys():
+        for annotated_example in data[intent]:
+            # Extract slot information
+            slot_annotations = re.findall(
+                r"\[(.*?)\]\((.*?)\)", annotated_example
+            )
+
+            # Remove slot annotations from the text
+            clean_text = re.sub(r"\[(.*?)\]\((.*?)\)", r"\1", annotated_example)
+
+            yield intent, clean_text, slot_annotations
+
+
+class JointBERTDataset(Dataset):
+    def __init__(self, path: str, max_length: int = 32):
+        self.data = load_yaml(path)
+        self.max_length = max_length
+
+        self.intent_label_count = len(JointBERTIntent)
+        self.slot_label_count = len(JointBERTSlot)
+
+        self.tokenizer = BertTokenizer.from_pretrained(_TOKENIZER_PATH)
+
+        self.examples = []
+        self._build_dataset()
+
+    def _build_dataset(self) -> None:
+        """Builds the dataset."""
+        for intent, clean_text, slot_annotations in parse_data(self.data):
+            intent, tokens, labels = self._tokenize_and_label(
+                intent, clean_text, slot_annotations
+            )
+
+            input_ids = self.tokenizer.encode(tokens, add_special_tokens=True)
+            attention_mask = [1] * len(input_ids)
+
+            # Add [CLS] and [SEP] tokens to labels
+            cls_label = _IGNORE_INDEX
+            sep_label = _IGNORE_INDEX
+            labels = [cls_label] + labels + [sep_label]
+
+            # Pad input_ids, attention_mask, and labels
+            padding_length = self.max_length - len(input_ids)
+            input_ids = input_ids + (
+                [self.tokenizer.pad_token_id] * padding_length
+            )
+            attention_mask = attention_mask + ([0] * padding_length)
+            labels = labels + ([_IGNORE_INDEX] * padding_length)
+            self.examples.append((input_ids, attention_mask, intent, labels))
+
+    def _num_word_tokens(self, word: str) -> int:
+        """Returns the number of word tokens in the input word.
+
+        Args:
+            word: The input word.
+
+        Returns:
+            The number of word tokens in the input word.
+        """
+        return len(self.tokenizer.tokenize(word))
+
+    def _tokenize_and_label(
+        self, intent: str, text: str, slot_annotations: Tuple(str, str)
+    ) -> Tuple[int, List[str], List[int]]:
+        """Tokenizes the text and assigns labels based on slot annotations.
+
+        Args:
+            intent: The intent of the text.
+            text: The text to tokenize.
+            slot_annotations: A tuple of slot-value pairs in the text.
+
+        Returns:
+            A tuple of the intent, tokenized text, and labels.
+        """
+        tokens = self.tokenizer.tokenize(text)
+        labels = []
+
+        start_idx = 0
+        for slot_text, slot_label in slot_annotations:
```

Comment:
Not sure if I follow what is going on here. A comment explaining the logic would be helpful. Maybe an example would help.

Reply:
Good point. Added example.